modules/skparagraph/src/TextWrapper.cpp

```diff
@@ -33,6 +33,90 @@ struct LineBreakerWithLittleRounding {
 
     const SkScalar fLower, fMaxWidth, fUpper;
 };
+
+struct HardLineBreakStyleHandler {
+    HardLineBreakStyleHandler(ParagraphImpl* parent, SkScalar maxWidth)
+        : fParent(parent)
+        , fMaxWidth(maxWidth)
+        , fHardLineBreakIndex(0)
+        , fTextAlign(parent->paragraphStyle().effective_align())
+        , fIndent(0)
+        , fExtraIndent(0)
+        , fUseInlineFontSizeForEmptyLine(false)
+        , fFirstLineOfHardBreakContext(true) {
+    }
+
+    // New hard line or soft line.
+    void newLineWillBegin(const Cluster* startCluster, const Cluster* endCluster) {
+        const std::vector<HardLineBreakStyle>& hardLineBreakStyles = fParent->paragraphStyle().getHardLineBreakStyles();
+        if (fHardLineBreakIndex >= hardLineBreakStyles.size()) {
+            // Should not happen. But if API is misused we reset everything.
+            resetAlignAndIndent();
+            return;
+        }
+
+        const HardLineBreakStyle& style = hardLineBreakStyles[fHardLineBreakIndex];
+        fTextAlign = fParent->paragraphStyle().effective_align(style.getTextAlign());
+        fIndent = style.getIndent();
+        fUseInlineFontSizeForEmptyLine = style.getUseInlineFontSizeForEmptyLine();
+        if (fFirstLineOfHardBreakContext) {
+            fFirstLineOfHardBreakContext = false;
+            fExtraIndent = 0;
+            // Look ahead to measure the width of the sticking text and record it in fExtraIndent.
+            // fExtraIndent is then used in the subsequent non-hard broken line: the 'else' branch.
+            if (size_t firstLineStickingTextLength = style.getFirstLineStickingTextLength()) {
+                size_t i = 0;
+                for (auto cluster = startCluster; cluster < endCluster && i < firstLineStickingTextLength; ++cluster, ++i) {
+                    fExtraIndent += cluster->width();
+                }
+            }
+        } else {
+            fIndent += fExtraIndent;
+        }
+    }
+
+    SkScalar getMaxWidthWithIndentSubtracted() const {
+        return std::max<SkScalar>(fMaxWidth - fIndent, 0);
+    }
+
+    SkScalar getIndent() const {
+        return fIndent;
+    }
+
+    TextAlign getTextAlign() const {
+        return fTextAlign;
+    }
+
+    bool getUseInlineFontSizeForEmptyLine() const {
+        return fUseInlineFontSizeForEmptyLine;
+    }
+
+    void hardLineBreak() {
+        ++fHardLineBreakIndex;
+        const std::vector<HardLineBreakStyle> hardLineBreakStyles = fParent->paragraphStyle().getHardLineBreakStyles();
+        if (fHardLineBreakIndex >= hardLineBreakStyles.size() || !hardLineBreakStyles[fHardLineBreakIndex].getIsFakeBreak()) {
+            fFirstLineOfHardBreakContext = true;
+            resetAlignAndIndent();
+        }
+    }
```

Comment:
1. fParent == null?
2.  please compare with line 52, Why are they any different?

Reply:
1. Not possible.
2. The difference is `fFirstLineOfHardBreakContext = true;`. So for normal `\n` I need to set it and reset align and indent while for fake one I don't need to since it is fake and it should continue with the current style.